### PR TITLE
[FSDP][1/N] `fully_shard` state dict

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90777 [FSDP][2/N] Refactor state dict hook registration
* **#90767 [FSDP][1/N] `fully_shard` state dict**
* #90660 [FSDP] Clean up `FlatParamHandle` dtypes, post-backward hook

Co-authored with @rohan-varma.

**Overview**
This adds preliminary `state_dict()` support for `fully_shard`.
- The only explicit branching between composable and wrapper code paths happens in the state dict hook registration, which is inevitable.
- We introduce a `_comm_module_prefix` to match the FQNs between the two code paths. This is needed since for composable, the FQNs are prefixed from the local FSDP root, whereas for state dict purposes, we want them to be prefixed from the comm. module. Thus, we need this `_comm_module_prefix` to be stripped during state dict.
    - In my understanding, the alternative to not use the `prefix` argument in `state_dict()` does not support the case when `fully_shard` is applied to a submodule (i.e. not the global root module) since we still need _part_ of `prefix` then.

**Follow-Ups**
- We can retire the `functools.partial` usage once @fegin's PR lands.
- We should add more thorough testing (e.g. sharded state dict, save and load together etc.).

### CODE REVIEW

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 1]
```diff
@@ -696,3 +704,57 @@ def _post_load_state_dict_hook(module: nn.Module, *args: Any) -> None:
...
+
+@no_type_check
+def _register_state_dict_hooks(state: _FSDPState) -> None:
+    """Registers the post-save state dict hooks."""
+    if not _is_composable(state):
+        state._register_state_dict_hook(functools.partial(_post_state_dict_hook, state))
+        return
+    for handle in state._handles:
+        handle._comm_module._register_state_dict_hook(
+            functools.partial(_post_state_dict_hook, state)
```
Comment (Skylion007) [inline torch/distributed/fsdp/_state_dict_utils.py:731]: Nit: shouldn't you cache these functools.partial calls? instead of regenerating every step in the loop?

Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:731]: Good point. I will only fix this if there are other changes requested to avoid re-running CI. We plan on removing this in a few days anyway.

Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:731]: Good point. I will only fix this if there are other changes requested to avoid re-running CI. We plan on removing this in a few days anyway.

File: test/distributed/_composable/test_fully_shard.py [hunk 2]
```diff
@@ -444,5 +444,53 @@ def test_float16_on_one_submodule(self):
...
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_root_fully_shard(self):
+        """
+        Tests that the full state dict saved from a module with ``fully_shard``
+        applied to the global root matches that of an equivalent local module.
+        """
+        local_model = CompositeParamModel(device=torch.device("cuda"))
+        composable_module = copy.deepcopy(local_model)
+        fully_shard(composable_module, policy=ModuleWrapPolicy({UnitModule}))
```
Comment (rohan-varma) [inline test/distributed/_composable/test_fully_shard.py:462]: is it worth checking class based, I guess maybe not, due to transitive nature since we've verified class based can be equivalent to local?

File: test/distributed/_composable/test_fully_shard.py [hunk 3]
```diff
@@ -444,5 +444,53 @@ def test_float16_on_one_submodule(self):
...
 
+class TestFSDPModelCheckpointing(FSDPTest):
+    """Tests composable FSDP model checkpointing."""
+
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_root_fully_shard(self):
```
Comment (rohan-varma) [inline test/distributed/_composable/test_fully_shard.py:455]: nit: might want to mention this is full state dict for now.

File: test/distributed/_composable/test_fully_shard.py [hunk 4]
```diff
@@ -444,5 +444,53 @@ def test_float16_on_one_submodule(self):
...
+
+    @skip_if_lt_x_gpu(2)
+    def test_state_dict_submodule_fully_shard(self):
+        """
+        Tests that the full state dict saved from a module with ``fully_shard``
+        applied on submodules matches that of an equivalent local module.
+        """
+        local_model = CompositeParamModel(device=torch.device("cuda"))
+        composable_module = copy.deepcopy(local_model)
+        fully_shard(composable_module.u1)
```
Comment (rohan-varma) [inline test/distributed/_composable/test_fully_shard.py:475]: as a follow up, we probably want to test manual wrapping when that is available as well. could we file an issue for this.

File: test/distributed/_composable/test_fully_shard.py [hunk 5]
```diff
@@ -444,5 +444,53 @@ def test_float16_on_one_submodule(self):
...
+    def _check_state_dict_parity(self, local_sd: Dict, composable_sd: Dict):
+        """Checks that ``local_sd`` and ``composable_sd`` are the same."""
+        # Check that all keys match
+        for k1, k2 in zip(composable_sd.keys(), local_sd.keys()):
+            self.assertEqual(k1, k2)
+        # Check that all values match
+        for v1, v2 in zip(composable_sd.values(), local_sd.values()):
+            self.assertEqual(v1.shape, v2.shape)
+        # Check that all keys and values are aligned
+        for (k1, v1), (k2, v2) in zip(composable_sd.items(), local_sd.items()):
```
Comment (rohan-varma) [inline test/distributed/_composable/test_fully_shard.py:490]: doesn't this encapsulate the above 2 checks?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 6]
```diff
@@ -696,3 +704,57 @@ def _post_load_state_dict_hook(module: nn.Module, *args: Any) -> None:
...
+        return
+    for handle in state._handles:
+        handle._comm_module.register_state_dict_pre_hook(
+            functools.partial(_pre_state_dict_hook, state)
+        )
+
+
+@no_type_check
+def _register_state_dict_hooks(state: _FSDPState) -> None:
+    """Registers the post-save state dict hooks."""
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:725]: to reduce repetition, might want to do 

```
hook = partial(_post_state_dict_hook, state)
```

and then reuse that for both cases.

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 7]
```diff
@@ -696,3 +704,57 @@ def _post_load_state_dict_hook(module: nn.Module, *args: Any) -> None:
...
+        )
+        return
+    for handle in state._handles:
+        handle._comm_module._register_load_state_dict_pre_hook(
+            functools.partial(_pre_load_state_dict_hook, state), with_module=True
+        )
+
+
+@no_type_check
+def _register_load_state_dict_post_hooks(state: _FSDPState) -> None:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:750]: I wonder if we can generalize all this to reduce code duplication:


```
def register_state_dict_hook(register_hook_fn, hook, state, hook_fn_kwargs):
    # same code for composable vs not composable
    getattr(state, register_hook_fn)(partial(hook, state)
```

and then pass in the different things into this function.

File: torch/distributed/fsdp/flat_param.py [hunk 8]
```diff
@@ -167,6 +167,9 @@ class FlatParameter(nn.Parameter):
             (i.e. some per-parameter state) used to customize pre-flatten and
             post-unflatten behavior. This is experimental, and users should not
             depend on its existence in the future.
+        _comm_module_prefix (str): Module name prefix starting from ``module``
+            to ``comm_module`` as passed to :class:`FlatParamHandle`, including
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:171]: One thing that I like to consider when writing comments is ease of understanding, i.e. would you expect an FSDP developer to clearly understand what "starting from module to comm_module" means in this context?

**Review by rohan-varma — APPROVED** [review]
LGTM overall, some minor comments inline. Thanks for hacking on this! 

Another related q: should we file issues to move `_unshard_utils` to be based entirely off the comm_module instead of _module_handles?

File: test/distributed/_composable/test_fully_shard.py (hunk 5 above)
Comment (awgu) [reply test/distributed/_composable/test_fully_shard.py:490]: Yes, but the error message will be better. I followed your unit test 😃

File: test/distributed/_composable/test_fully_shard.py (hunk 2 above)
Comment (awgu) [reply test/distributed/_composable/test_fully_shard.py:462]: Yes, I think our goal is to match local, so we can test that directly. The transitivity holds as well.

**Comment by awgu** [comment]
> Another related q: should we file issues to move `_unshard_utils` to be based entirely off the comm_module instead of _module_handles?

The _cool_ thing about this implementation is that `_module_handles` gets the handle from `_comm_module_to_handles`, so there is no need to use `_module_to_handles`. I need to double check, but we may be able to remove `_module_to_handles` now.

File: torch/distributed/fsdp/flat_param.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:171]: Done in follow-up.

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:750]: Done in follow-up.

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 6 above)
Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:725]: Done in follow-up.

File: test/distributed/_composable/test_fully_shard.py (hunk 3 above)
Comment (awgu) [reply test/distributed/_composable/test_fully_shard.py:455]: In the line below:
"Tests that the full state dict"

File: torch/distributed/fsdp/flat_param.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/fsdp/flat_param.py:171]: I think that an FSDP developer should understand this. The design of FSDP is complex enough where it is important for developers to understand the difference between `module` and `comm_module`, or else it is hard to follow the differences between wrapper FSDP and `fully_shard`.

I try to always document what every variable means and the difference for wrapper vs. composable code paths.

**Comment by awgu** [comment]
The single failure looks unrelated:
```
= 1 failed, 2114 passed, 858 skipped, 17 deselected, 28 xfailed, 1 warning, 2 rerun in 500.98s (0:08:20) =
```

```
____________ TestCommonCUDA.test_out_warning_linalg_lu_factor_cuda _____________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 757, in test_out_warning
    _compare_out(_case_zero_transform, compare_strides_and_data_ptrs=False)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 1317, in wrapper
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 731, in _compare_out
    op_out(out=out)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 1068, in __call__
    return self.op(*args, **kwargs)
RuntimeError: torch.linalg.lu_factor: (Batch element 0): U[1,1] is zero and using it on lu_solve would result in a division by zero. If you still want to perform the factorization, consider calling linalg.lu(A, pivot) or linalg.lu_factor_ex(A, pivot)
____________ TestCommonCUDA.test_out_warning_linalg_lu_factor_cuda _____________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 676, in test_out_warning
    expected = op(sample.input, *sample.args, **sample.kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 1068, in __call__
    return self.op(*args, **kwargs)
RuntimeError: torch.linalg.lu_factor: U[1,1] is zero and using it on lu_solve would result in a division by zero. If you still want to perform the factorization, consider calling linalg.lu(A, pivot) or linalg.lu_factor_ex(A, pivot)
=================================== FAILURES ===================================
____________ TestCommonCUDA.test_out_warning_linalg_lu_factor_cuda _____________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_ops.py", line 676, in test_out_warning
    expected = op(sample.input, *sample.args, **sample.kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/opinfo/core.py", line 1068, in __call__
    return self.op(*args, **kwargs)
RuntimeError: torch.linalg.lu_factor: U[1,1] is zero and using it on lu_solve would result in a division by zero. If you still want to perform the factorization, consider calling linalg.lu(A, pivot) or linalg.lu_factor_ex(A, pivot)
```

File: torch/distributed/fsdp/_common_utils.py [hunk 9]
```diff
@@ -68,7 +68,10 @@ def _module_handles(state: _FSDPState, module: nn.Module) -> List:
     module is FullyShardedDataParallel, the module._handles will be returned.
     """
     if _is_composable(state):
-        return state._module_to_handles[module][:]
+        assert (
+            module in state._comm_module_to_handles
+        ), f"Expects a `comm_module` but got {module} on rank {state.rank}"
+        return state._comm_module_to_handles[module][:]
```
Comment (fegin) [inline torch/distributed/fsdp/_common_utils.py:74]: Thanks for fixing this. optimizer state_dict will use this API too.

Comment (fegin) [reply torch/distributed/fsdp/_common_utils.py:74]: Will module wrapper FSDP also switch to `_comm_modules` in the future?

**Review by fegin — APPROVED** [review]
One general question: are we going to separate the state_dict tests for composable FSDP and module wrapper FSDP or do you think it is valuable to consolidate them into one? Asking this because I'm also writing one for optimizer state_dict.

File: torch/distributed/fsdp/_common_utils.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/fsdp/_common_utils.py:74]: > Will module wrapper FSDP also switch to `_comm_modules` in the future?

I think there is no need for wrapper FSDP. In the current design, each `_comm_module` is exactly the `FullyShardedDataParallel._fsdp_wrapped_module`, which is basically the same thing as the `FullyShardedDataParallel` instance itself, so `module._handles[:]` should be correct.

**Comment by awgu** [comment]
> One general question: are we going to separate the state_dict tests for composable FSDP and module wrapper FSDP or do you think it is valuable to consolidate them into one?

Chatting briefly offline with @rohan-varma, I think we can write some basic tests for composable path and take advantage of the fact that the two paths share underlying subroutines to avoid too much duplication. Ideally, we can consolidate both test suites, but I am not sure if that will be feasible from an engineering perspective. We can still set that as the goal.

@rohan-varma  feel free to provide your opinion here. I do not feel too strongly.

**Comment by awgu** [comment]
linux-focal-rocm5.2-py3.8 / test (default, 1, 2, linux.rocm.gpu) time out looks unrelated.
